training_config/bidirectional_lm.jsonnet

```diff
@@ -0,0 +1,152 @@
+local NUM_GPUS = 2;
+local NUM_THREADS = 1;
+
+local BASE_READER = {
+        "type": "simple_language_modeling",
+        "tokenizer": {
+          "type": "word",
+          "word_splitter": {
+	        # The 1 Billion Word Language Model Benchmark dataset is
+	        # pre-tokenized. (Also, if you're running against a untokenized
+	        # dataset be aware that there are serialization issues with Spacy.
+	        # These come into play in the multiprocess case.)
+            "type": "just_spaces"
+          }
+        },
+        "token_indexers": {
+          "tokens": {
+            "type": "single_id"
+          },
+          "token_characters": {
+            "type": "elmo_characters"
+          }
+        },
+        "max_sequence_length": 500
+};
+
+local BASE_ITERATOR = {
+  "type": "bucket",
+  "max_instances_in_memory": 16384 * NUM_GPUS,
+  # Larger than we really desire for a batch. Since we set
+  # maximum_samples_per_batch below we will pack approximately that many
+  # samples in every batch.
+  "batch_size": 512 * NUM_GPUS,
+  "sorting_keys": [["source", "num_tokens"]],
+  "maximum_samples_per_batch": ["num_tokens", NUM_GPUS * 1000]
+};
+
+{
+  "dataset_reader": {
+    "type": "multiprocess",
+    "base_reader": BASE_READER,
+    "num_workers": NUM_THREADS,
+    "output_queue_size": 1000
+  },
+  # Note: We don't set a validation_data_path because the softmax is only
+  # sampled during training. Not sampling on GPUs results in a certain OOM
+  # given our large vocabulary. We'll need to evaluate against the test set
+  # (when we'll want a full softmax) with the CPU.
+  "train_data_path": "/home/brendanr/workbenches/calypso/train/*",
```

Comment:
Could you pull this up into a env variable so that anyone could use this config without having to modify it?

Reply:
Done. Should we host the 1b corpus on S3 as well? It's 1.7GB compressed. Also, since it's in multiple parts we might need to change the code that manages tarred inputs. Haven't looked at that code yet.